### DISCUSSION TITLE
Emanote and Neovim

### DISCUSSION THREAD

**Discussion by j-steinbach** [srid/emanote#237] [General]
We should collect all Neovim (and base Vim) related things in one place, so that we can compile it down to a template, similar to the  VS Code template ([here](https://github.com/srid/emanote-template) and [here](https://emanote.srid.ca/resources/emanote-template))

----

Based on the talk https://github.com/srid/emanote/issues/180.

**Comment by srid** [comment]
Good idea. Would be happy to help integrate that into `emanote-template`. 

Ideally, we want to use Nix for reproducibility. But I'm happy even if we are just adding the Vim configuration files in the repo along with manual instructions in README.

I don't use Vim, so a PR is appreciated.

**Reply by srid** [reply]
Yes, there is `.vscode/` - but the repo is not necessarily tied to VSCode. We can have multiple editor configurations.

**Reply by j-steinbach** [reply]
Personally I don't like or use Nix (or Home Manager) for my configs/dotfiles, so from my side I can only provide config files. But let's postpone this discussion until I (or someone else) actually posts some config files :)

**Comment by j-steinbach** [comment]
I am still working on learning and configuring Neovim, but I want to list some findings. I intend this post to be an open-ended overview over Vim, mostly focused on note-taking, without diving too much into configuration details. Technical details are simplified and I am not (yet) providing config examples. I likely will follow up with or link my own personal and opinionated config some time in the future (a.k.a. _never_.)

## Getting started with Vim

[Learn Vim](https://github.com/iggredible/Learn-Vim) is an amazing resource. You can also use the in-built `:Tutor` to get a quick peek at Vim's capabilities.

Vim or Neovim? [Neovim](https://neovim.io/). Basically Neovim is Vim, except with a modern code-base. That means speed, reliability and new features. Whenever I talk about Vim I actually mean Neovim :)

## Configuring Vim

This is a rabbit-hole. To just get started create a `~/.config/nvim/init.vim` file, then add your settings there. (How to actually do that is out-of-scope, but you _will_ find tutorials if you search for `how to configure neovim`). Still, you probably want to disable backwards-compatibility (`:set nocompat`) and enable more colors in the termianl (`:set termguicolors`).

For more advanced users you can also start with a `init.lua`. See the [nanotee guide](https://github.com/nanotee/nvim-lua-guide) for more information. 

## Installing Plugins

There are a bunch of different plugin-managers for Vim. I use [packer.nvim](https://github.com/wbthomason/packer.nvim)

## Vim and Markdown

Emanote does the heavy lifting for rendering and displaying my [Zettelkasten](https://zettelkasten.de/), but I still need a way to create and edit my notes. This is where Vim comes into play. I believe it all boils down to the following minimal features:

- markdown language support (syntax-highlighting, helpful shortcuts, ...)
- wiki-functionality (following links, auto-completion of links, ...)

### Wiki functionality (unsorted)

A bunch of options that should provide the needed wiki functionality.

- [follow-md-links.nvim](https://github.com/jghauser/follow-md-links.nvim)
- [mkdx](https://github.com/SidOfc/mkdx) 
- [mkdnflow.nvim](https://github.com/jakewvincent/mkdnflow.nvim)
- [vim-markdown](https://github.com/plasticboy/vim-markdown) _(also does syntax-highlighting)_
    - [nvim-arkdown](https://github.com/ixru/nvim-markdown) _(also does syntax-highlighting)_
- [telekasten.nvim](https://github.com/renerocksai/telekasten.nvim) -- the option I am currently evaluating
- [vimwiki](https://github.com/vimwiki/vimwiki) -- can be used with markdown syntax
- [wiki.vim](https://github.com/lervag/wiki.vim)

There are also tools like [zk](https://github.com/mickael-menu/zk) that provide Vim integration. And I want to mention [nb](https://github.com/xwmx/nb), because it is cool.

### Markdown syntax highlight

Chances are that you want to highlight syntax your markdown in the editor. For example display links in a different color or make the headings bold. This is called syntax-highlighting and there are two ways to do that -- RegEx or Tree-sitter.

#### RegEx

[RegEx](https://www.wikiwand.com/en/Regular_expression) (Regular Expression) is the old and tested way. The computer goes through the document and highlights every match that fits the `regex`. The problem: Markdown is very difficult to pars and RegEx is slow.

- [vim-markdown](https://github.com/plasticboy/vim-markdown)
    - [nvim-markdown](https://github.com/ixru/nvim-markdown)

#### Tree-sitter

[Tree-sitter](https://tree-sitter.github.io/tree-sitter/) is the new kid on the block. Faster, better, harder, stronger. (Don't ask me for details)

- [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter)
- [tree-sitter-markdown](https://github.com/ikatyang/tree-sitter-markdown)

### LSP (Advanced)

[LSP](https://microsoft.github.io/language-server-protocol/) is a way to add IDE features to non-IDE editors, like [Neovim](https://neovim.io/doc/lsp/). That means Neovim will highlight syntax-errors in your Markdown document. Still, this is more of a _nice-to-have_ feature.

- [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig)
- [remark LSP](https://github.com/remarkjs/remark-language-server)

## Other Plugins

There are a few other plugins I want to entrust you with.

### Finding Files

As mentioned in [Learn Vim](https://github.com/iggredible/Learn-Vim/blob/master/ch03_searching_files.md), a (fuzzy) file finder is one of the biggest improvements you can make in your day-to-day use. I can't believe I did that manually!

- [fzf.vim](https://github.com/junegunn/fzf.vim) -- the fastest kid on the block
- [telescope.nvim](https://github.com/nvim-telescope/telescope.nvim) -- a little less fast, but with more QOL. Just don't forget to install the [suggested dependencies](https://github.com/nvim-telescope/telescope.nvim#suggested-dependencies).

### Distraction-free writing / Zen Mode

I often write longer texts (like this one) or novels and really need a _no distraction_ mode. I also recommend to watch [this](https://www.youtube.com/watch?v=oMG7Cuc3v7A&list=PLHmevVAAXtu3_beDLtsTmkivMI__2b6P8&index=2) or read [Csikszentmihalyi's Theory of Flow](https://link.springer.com/chapter/10.1007/978-94-017-9088-8_16).

- [goyo.vim](https://github.com/junegunn/goyo.vim) _and_ [limelight.vim](https://github.com/junegunn/limelight.vim)

### WhichKey

Vim is complicated and has hundreds of shortcuts. Wouldn't it be great if someone or something could give me a quick overview whenever I press a key? 

- [which-key.nvim](https://github.com/folke/which-key.nvim)

### Cosmetics

Plain Vim is ugly as hell, but it doesn't have to be this way.

For the _colour scheme_ I use either [everforest](https://github.com/sainnhe/everforest) or [embark](https://github.com/embark-theme/vim), but you can find more [here](https://vimcolorschemes.com/) or [here](https://github.com/rafi/awesome-vim-colorschemes) or [here](https://github.com/flazz/vim-colorschemes).

For the _statusline_ I currently use [feline.nvim](https://github.com/feline-nvim/feline.nvim).

## More plugins

You can find more plugins on [vimawesome](https://vimawesome.com/) and [neovimcraft](https://neovimcraft.com/) and. Or just search on the web, d'oh.

**Reply by edrex** [reply]
I'm not currently using vim for wiki editing, but have in the past and will likely again. @lervag's wiki.vim is pretty good. While there's a temptation to recommend a full set of plugins and config, it may be most generally useful to focus on config that directly enables wiki editing functions, so folks can adapt to their own config preferences, possibly with a shout to one of the base configs such as https://github.com/liuchengxu/space-vim

**Reply by j-steinbach** [reply]
[SpaceVim](https://github.com/SpaceVim/SpaceVim) is a good point. Maybe we should first discuss for whom we want to make a guide for.

I see four kinds of possible groups:

- People that are no-tech and just want to get started taking notes. I don't think this group should get into Vim
- People that have no experience with Vim but are willing to dive into it and get their hands dirty. They likely need an opinionated guide with training wheels.
- People that are already using Vim, but not for long and maybe not for note-taking. (I count my self in this group).
- People that live and breathe Vim. They won't need a guide, but maybe some inspiration. 

In my opinion this narrows down to the 2nd and the 3rd group, with the 2nd one being the main one.

**Reply by edrex** [reply]
liuchengxu/space-vim isn't the same as SpaceVim/SpaceVim. I found the latter a bit of a dumping ground, while the former is a focused and efficient product of a single developer.

**Reply by srid** [reply]
Fifth kind of group,

- People who extend neovim in Haskell

Using https://github.com/tek/ribosome - I imagine we can do a more direct integration with Emanote.

**Reply by j-steinbach** [reply]
@edrex you seem to know a bit about this. Would you mind just listing the Vim "frameworks" you know? And maybe add some commentary? I just went by name and looked at the one with the biggest number of stars.

**Comment by Ninlives** [comment]
Are there any vim/nvim plugin that supports the emanote style wiki link now (syntax highlighting, follow links, etc..)?

**Reply by j-steinbach** [reply]
What exactly do you mean with " supports emanote style wiki link"? Can you give an example for the syntax?

**Reply by Ninlives** [reply]
For example emnanote support wiki link in this for `[[neuron|Moving off neuron]]` where `neuron` is the file name and `Moving off neuron` is the custom title. I don't know if there is any plugin that respect this feature. Another problem is for the resolving. In emanote `[[neuron]]` will be resolved to a file under to root directory whose name is 'neuron', no matter where it is placed (as far as there is only one file named 'neuron'), but the resolving rule seems to be different for different plugins.

**Reply by j-steinbach** [reply]
Just checked it for Telekasten, which is what I use. It searches for the full string, i.e. a note named `neuron|Moving off neuron` :/ I guess this is the problem with all the markdown flavours and dialects -- no uniformed syntax. Telekasten however searches in the full root path with fuzzy finding, which means it also detects something `root/foo/bar/baz/mynote.md`. If I have time I will also try Vimwiki, but I doubt I can give a 100% positive response.

**Reply by applejag** [reply]
I've started some docs on using `zk`, which can be configured to support Emanote/Neuron's wiki link style: https://github.com/EmaApps/emanote/pull/321/files?short_path=b7fc022#diff-b7fc0227b5af24019e96ab156468683c90efd4e25dcaf56c9be3b3af9415718f

E.g:

```toml
# .zk/config.toml

[format.markdown]

# Given "content/mydir/filename.md", will render "[[filename]]"
link-format = "[[{{filename}}]]"
```

`zk` can be added as a LSP server so it will autocomplete notes and write the correct `[[filename]]` links on completion.
